[16.0][MIG] rma_sale: Migration to 16.0

pedrobaeza opened the pull request:
- Standard procedure.
- Adapt to new `prepare_*_vals` methods.
- Renamed fields adaptation.

@Tecnativa TT44214

Commit:
[ADD] rma_sale: new module

[UPD] Update rma_sale.pot

[ADD] icon.png

fcolus committed:
Added translation using Weblate (Portuguese (Brazil))

Commit:
[FIX] rma_sale: fix expected single record error in tests.

rma_sale 12.0.1.0.1

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/

chienandalu committed:
[FIX+IMP] rma: views permissions + portal permissions + teams flow

[FIX] rma: views permissions

Regular users don't have permissions to rma models, so we should avoid
loading views that lead to permission errors.

TT24986

rma_sale 12.0.1.1.0

[FIX] rma_sale: portal permissions

Some operations need to be sudoed to be reachable by the portal user

[UPD] Update rma_sale.pot

rma_sale 12.0.1.1.1

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/

[IMP] rma: teams flow

- If no RMA Team is set, we'll assign a default one to the new RMA.
- A sequence is now used to search for the top team and assign it.
- No default user is assigned when it's not in the context (i.e. portal
rmas).

[UPD] Update rma_sale.pot

rma_sale 12.0.1.2.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/

Commit:
[FIX+IMP] rma, rma_sale: fix bugs and add improvements

- Fix thrown error when trying to download a picking from the portal.
- Add the hook method to prepare RMA values ​​from the return pick wizard.
- Add the access rule for portal users.
- Show the portal 'Request RMAs' button on the sales page only to users
related to the sales order.

[UPD] Update rma_sale.pot

rma_sale 12.0.1.3.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/

pedrobaeza committed:
Translated using Weblate (Spanish)

Currently translated at 81.5% (44 of 54 strings)

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/es/

chienandalu committed:
[IMP] rma_sale: portal improvents

- Now it's possible to open several RMAs in a sale order from the portal
- A new comment button has been added to allow the portal user to enter
relevant information like serial numbers o issue description.
- If the requested operation isn't set no RMA will be opened
- The RMA product qty is now a numeric control with limits according to
the qty available to return

[FIX] rma,rma_sale: fix linter errors

[UPD] Update rma_sale.pot

rma_sale 12.0.1.4.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/

pedrobaeza committed:
[FIX+IMP] rma_sale: Proper dependency + propagate salesman

- Proper dependency is `sale_stock`, not `sale`, as we are using some fields added
  by this module.
- Propagate salesman from sales order when available.

TT25525

rma_sale 12.0.1.4.1

chienandalu committed:
[FIX+IMP] rma: usability + Translated using Weblate (Spanish)

[FIX+IMP] rma: usability

* IMP - Now the description will be an html son we can show rich styles
in the customers report.
* FIX - On locked sale orders it was need to unlock them to be able to open an RMA.
* IMP - Make the description label visible in the backend form so the
user can easily spot it.
* IMP - Added date and deadline filters.
* IMP - Added pending RMAs filter.
* IMP - Added late RMAs filter.
* IMP - Added danger decoration in tree view

rma_sale 12.0.1.5.0

Translated using Weblate (Spanish)

Currently translated at 100.0% (62 of 62 strings)

Translation: rma-12.0/rma-12.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-12-0/rma-12-0-rma_sale/es/

Commit:
[FIX] rma_sale: show related RMAs without default filters

rma_sale 12.0.1.5.1

pedrobaeza committed:
[FIX] rma_sale: Show directly form on 1 RMA

Fine-tune of 9a25d6, as for one record, the domain is not applied, but the list is
shown.

Commit:
[IMP] rma_sale: black, isort, prettier

Commit:
[MIG] rma_sale: Migration to 13.0

chienandalu committed:
[IMP] rma, rma_sale: shipping address

Allow to set the desired shipping address where the goods must be
returned after the RMA is processed.

chienandalu committed:
[IMP] rma_sale: prepare kits integration

When a sale line has a phantom product (mrp kits) the RMA would not be
possible as the wizard couldn't pair the components moves with the
product in the line. With this approach, we can at least return the
spare components of the original kit line.

We also need some hooks to intervine in the main methods, like in
invoicing.

chafique-delli committed:
[14.0][MIG] rma_sale module

Commit:
[FIX] rma_sale: Request RMA from SO without picking from portal

dhongu committed:
Added translation using Weblate (Romanian)

Currently translated at 100.0% (63 of 63 strings)

Translation: rma-13.0/rma-13.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-13-0/rma-13-0-rma_sale/ro/

Commit:
[IMP] *rma*: Avoid RMAs of kits until rma_mrp is installed

Commit:
[IMP] *rma*: some improvements

bosd committed:
Added translation using Weblate (Dutch)

Currently translated at 61.9% (39 of 63 strings)

Translation: rma-13.0/rma-13.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-13-0/rma-13-0-rma_sale/nl/

chienandalu committed:
[FIX] rma_sale: product in multiple lines

If there are several stock moves with the same product in the picking we
won't be able to make the RMA

chienandalu committed:
[IMP] rma_sale: portal ux

When the user selects an operation, the comment shows up so the customer
doesn't forget to enter his comments

SilvioGregorini committed:
[FIX] rma_sale: fix wrong values for RMA creation

Lesser fix for creating RMA from website.
Current values will break any custom behaviour because Many2one fields are set using strings instead of integer (IDs).

SilvioGregorini committed:
[FIX] rma_sale: fix CacheError in _prepare_rma_values()

Lesser fix for sale.order.line.rma.wizard methods.
When method '_prepare_rma_values' was called upon records where field 'picking_id' was empty, Odoo raised a CacheError when trying to access field 'move_id'.
That happened because computed method '_compute_move_id' was not assigning a proper value to such field when 'picking_id' was empty.
Once the computed method is fixed (by simply assigning 'False' as 'move_id' value when no picking is set), the CacheError is solved.

chienandalu committed:
[FIX] rma_sale: recurrent RMA

If a product was already in an RMA in the past, we should be able to
place another RMA in the future.

chienandalu committed:
[FIX] rma_sale: view id

chienandalu committed:
[FIX] rma_sale: take product from move

chienandalu committed:
[FIX] rma_sale: chained returns

Using move_dest_ids we can easily end in an infinite loop situation as
the return of the return of the return ends with some original moves on
in move_dest_ids. We must ensure to drop them to avoid the infinite
loop.

TT29886

pedrobaeza committed:
[FIX] rma: pre-commit

chienandalu committed:
[IMP] rma_sale: portal rma request single page view

Now it's possible to configure if the portal RMA request form is loaded
in a popup or in a single page.

In that page, we can add custom blocks (if the website is installed) a
customize the form text.

In this commit, we also add the possibility to extend the form view to allow
custom fields that will show up in the RMA description.

TT29670

chafique-delli committed:
[IMP] rma_sale: black, isort, prettier

sebastienbeau committed:
[FIX] fix filter rma location based on the sale order company

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

Currently translated at 100.0% (79 of 79 strings)

Translation: rma-14.0/rma-14.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-14-0/rma-14-0-rma_sale/pt/

francesco-ooops committed:
Added translation using Weblate (Italian)

chienandalu committed:
[IMP] rma_sale: avoid submit form in portal if no info

TT33572

chienandalu committed:
[FIX] rma_sale: subscribe portal users to RMA notifications

TT35269

SicurSam committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (79 of 79 strings)

Translation: rma-14.0/rma-14.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-14-0/rma-14-0-rma_sale/it/

GSLabIt committed:
[FIX] rma_sale: issue opening rma view from wizard + action permission issues

victoralmau committed:
[FIX] rma + rma_sale: Allow to create an RMA to a user with access_to…ken to sale order (no user created).
[FIX] rma + rma_sale: Show RMA's list (with share url) to a user with access_token to sale order (no user created).

Alex-Vining committed:
[MIG] rma_sale: Migration to 15.0

victoralmau committed:
[FIX] rma_sale: Filter done pickings and done moves

TT39089

victoralmau committed:
[FIX] rma_sale: Set picking_id field to required in wizard to avoid creating
rma if nothing has been delivered yet

TT39089

victoralmau committed:
[FIX] rma_sale: Allow to request a rma from portal order page only with portal user permission

victoralmau committed:
[IMP] rma_sale: Improve tests

TT37303

pedrobaeza committed:
[IMP] rma_sale: Domains performance

For avoiding a big list of IDs being transferred when no sales order
is selected on the RMA, we have changed domains to make use of the
possibility of pyjs expressions allowed in the domains.

No ternary operators nor list sums are allowed in pyjs, but using a
combination of allowed IDs with a controlled length of values + and/or
operators to switch domains is enough for having the right performance
and avoid to depend on other modules like web_domain_field.

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (79 of 79 strings)

Translation: rma-15.0/rma-15.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-15-0/rma-15-0-rma_sale/it/

pedrobaeza committed:
[FIX+IMP] rma_sale: Link invoice/move line with origin sale line when refunding

Steps to reproduce:

- Create a sales order with an storable product with invoicing policy
  on delivered quantities.
- Confirm it and deliver the product.
- Invoice the order.
- Do an RMA, receive it, and refund it.

Result: the delivered quantity is 1 instead of 0.

This is because the refund generated from the RMA is not linked to
sales order line, nor the RMA reception move. This is done because
other operations are performed:

- Be replaced.
- Be changed by other product.

And we don't also want that meanwhile the RMA is being performed, the
sales order is pending to invoice.

But when the refund has been done, we have it clear, so let's link both
and have sales statistics correct.

FIX: We don't link the refund line with the sales order if the RMA
quantity is not the whole original move quantity. Otherwise, we will
have incoherente delivered/invoiced quantities on the sales order.

TT41645

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (79 of 79 strings)

Translation: rma-15.0/rma-15.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-15-0/rma-15-0-rma_sale/it/

Commit:
Added translation using Weblate (German)

Currently translated at 100.0% (79 of 79 strings)

Translation: rma-15.0/rma-15.0-rma_sale
Translate-URL: https://translation.odoo-community.org/projects/rma-15-0/rma-15-0-rma_sale/de/

Commit:
[UPD] Update rma_sale.pot

chienandalu committed:
[IMP] rma_sale: rollback sale refund on deleted invoice

TT44043

pedrobaeza committed:
[IMP] rma_sale: pre-commit stuff

pedrobaeza committed:
[MIG] rma_sale: Migration to 16.0

- Standard procedure.
- Adapt to new `prepare_*_vals` methods.
- Renamed fields adaptation.
- Improved test coverage.

TT44214

s

pedrobaeza commented:
/ocabot migration rma_sale

chienandalu reviewed (changes requested):
Backend functionality ok. Some issues I came into:

The wizard form window doesn't show up properly

![image](https://github.com/OCA/rma/assets/5040182/31f97869-d287-4e0d-bdd0-d89b1a98963f)

Looks like the portal wizard backend js code needs to be retailored as it doesn't work properly. Do you want me to take a look into it?

pedrobaeza commented:
Ouch, I looked for o2m in views, but not in wizards. Fixed.

About the portal JS, please do the needed changes, and it would be good to have a tour test for checking it.

chienandalu commented on rma_sale/views/sale_portal_template.xml:
```diff
@@ -0,0 +1,327 @@
+                <button
+                    class="btn btn-primary btn-block mb8"
+                    type="button"
+                    data-toggle="collapse"
+                    data-target="#delivery_address_picker"
```
```suggestion
                    data-bs-toggle="collapse"
                    data-bs-target="#delivery_address_picker"
```

chienandalu commented on rma_sale/views/sale_portal_template.xml:
```diff
@@ -0,0 +1,327 @@
+                    aria-expanded="false"
+                ><i class="fa fa-truck" /> Choose a delivery address</button>
+                <div class="col-lg-12 collapse mt8" id="delivery_address_picker">
+                    <div data-toggle="buttons" class="row">
```
```suggestion
                    <div data-bs-toggle="buttons" class="row">
```

chienandalu commented on rma_sale/views/sale_portal_template.xml:
```diff
@@ -0,0 +1,327 @@
+                                        <button
+                                            class="btn btn-primary fa fa-comments"
+                                            type="button"
+                                            data-toggle="collapse"
```
```suggestion
                                            data-bs-toggle="collapse"
```

chienandalu commented on rma_sale/views/sale_portal_template.xml:
```diff
@@ -0,0 +1,327 @@
+                        role="button"
+                        class="btn btn-secondary btn-block mb8"
+                        data-toggle="modal"
+                        data-target="#modal-request-rma"
```
```suggestion
                        data-bs-toggle="modal"
                        data-bs-target="#modal-request-rma"
```

chienandalu reviewed (changes requested):
Finally no js revamp was needed but just some bs5 attributes tweaking

chienandalu committed:
[MIG] rma_sale: sale portal

- Migrate bs4 to bs5
- Fix shipping address choice functionality
- Add tour for the user portal workflow

chienandalu commented:
I just tried this PR https://github.com/OCA/openupgradelib/pull/338 to migrate the qweb templates and it looks that its working nicely :) I added a commit to the PR

chienandalu commented on rma_sale/views/sale_portal_template.xml:
```diff
@@ -61,15 +61,15 @@
                                 name="partner_shipping_id"
                                 t-att-value="address.id"
-                            >
+                            />
```
<strike>This change is weird. I guess, the bs4>bs5 scripts need some tweaking :) I'll fix it by hand</strike>

<strike>Ah, no, indeed it's correct to close it...</strike>

I'll check functionally

chienandalu replied:
Ok, now I remember why this was strange. This was a bs4 hack to select the rma address from the address cards with no js. I'll check how to adapt it to make it work, but the change was correct, I think.

pedrobaeza commented:
Is it correct then? Are you going to prepare a tour test?

chienandalu commented:
> Is it correct then? Are you going to prepare a tour test?

No, I should fix the delivery address selector in the portal. I can prepare a tour, yes.

chienandalu reviewed (approved):
- portal functionality fixed
- tours added

pedrobaeza commented:
Is this finished then?

chienandalu commented:
Yes :)

pedrobaeza commented:
I have improved a bit more the tests coverage and fine-tune several warnings, so it's time to merge it:

/ocabot merge nobump
